File: go/mechanisms/svm/upto/facilitator/signer.go
package facilitator

import (
	"context"
	"fmt"

	solana "github.com/gagliardetto/solana-go"
	"github.com/gagliardetto/solana-go/rpc"

	"github.com/x402-foundation/x402/go/v2/mechanisms/svm"
	"github.com/x402-foundation/x402/go/v2/mechanisms/svm/paymentchannels"
)

// UptoFacilitatorSigner is FacilitatorSvmSigner narrowed to the read and
// simulate RPC the `upto` facilitator requires at runtime. Exact-only signers
// omit the extra methods.
type UptoFacilitatorSigner interface {
	svm.FacilitatorSvmSigner

	GetAccountInfo(
		ctx context.Context,
		account solana.PublicKey,
		network string,
		opts *rpc.GetAccountInfoOpts,
	) (*rpc.GetAccountInfoResult, error)
	GetLatestBlockhash(ctx context.Context, network string) (solana.Hash, uint64, error)
	GetSlot(ctx context.Context, network string, commitment rpc.CommitmentType) (uint64, error)
	SimulateTransactionWithOpts(
		ctx context.Context,
		tx *solana.Transaction,
		network string,
		opts *rpc.SimulateTransactionOpts,
	) error
	GetProgramAccounts(
		ctx context.Context,
		network string,
		programID solana.PublicKey,
		opts *rpc.GetProgramAccountsOpts,
	) (rpc.GetProgramAccountsResult, error)
}

// assertUptoFacilitatorSigner validates that signer exposes every RPC cap the
// `upto` facilitator needs at construction. Panics with a clear message when a
// required method is missing.
func assertUptoFacilitatorSigner(signer svm.FacilitatorSvmSigner, label string) UptoFacilitatorSigner {
	type accountInfoGetter interface {
		GetAccountInfo(context.Context, solana.PublicKey, string, *rpc.GetAccountInfoOpts) (*rpc.GetAccountInfoResult, error)
	}
	type blockhashGetter interface {
		GetLatestBlockhash(context.Context, string) (solana.Hash, uint64, error)
	}
	type slotGetter interface {
		GetSlot(context.Context, string, rpc.CommitmentType) (uint64, error)
	}
	type simWithOpts interface {
		SimulateTransactionWithOpts(context.Context, *solana.Transaction, string, *rpc.SimulateTransactionOpts) error
	}

	if _, ok := signer.(accountInfoGetter); !ok {
		panic(fmt.Sprintf("%s requires GetAccountInfo on the signer", label))
	}
	if _, ok := signer.(blockhashGetter); !ok {
		panic(fmt.Sprintf("%s requires GetLatestBlockhash on the signer", label))
	}
	if _, ok := signer.(slotGetter); !ok {
		panic(fmt.Sprintf("%s requires GetSlot on the signer", label))
	}
	if _, ok := signer.(simWithOpts); !ok {
		panic(fmt.Sprintf("%s requires SimulateTransactionWithOpts on the signer", label))
	}
	return signer.(UptoFacilitatorSigner)
}

// assertProgramAccountsSigner validates GetProgramAccounts for discovery sweeps.
func assertProgramAccountsSigner(signer svm.FacilitatorSvmSigner, label string) UptoFacilitatorSigner {
	uptoSigner := assertUptoFacilitatorSigner(signer, label)
	type programAccountsGetter interface {
		GetProgramAccounts(context.Context, string, solana.PublicKey, *rpc.GetProgramAccountsOpts) (rpc.GetProgramAccountsResult, error)
	}
	if _, ok := signer.(programAccountsGetter); !ok {
		panic(fmt.Sprintf("%s requires GetProgramAccounts on the signer", label))
	}
	return uptoSigner
}

type programAccountsQuerier struct {
	signer  UptoFacilitatorSigner
	network string
}

func (q programAccountsQuerier) GetProgramAccounts(
	ctx context.Context,
	opts *rpc.GetProgramAccountsOpts,
) (rpc.GetProgramAccountsResult, error) {
	return q.signer.GetProgramAccounts(ctx, q.network, paymentchannels.ProgramID, opts)
}

File: typescript/packages/mechanisms/svm/src/payment-channels/signer.ts
import type { Address } from "@solana/kit";

import type { FacilitatorSigningCapabilities, FacilitatorSvmSigner } from "../signer";
import { fetchMaybeChannel } from "./generated/accounts/channel";

/**
 * {@link FacilitatorSvmSigner} narrowed to the optional caps payment-channel
 * facilitator work requires: reading a channel, a slot, and a blockhash, and
 * resolving a kit signer. Exact-only signers omit these methods, so they stay
 * optional on the base type. `upto` and batch settlement both use this set.
 */
export type PaymentChannelFacilitatorSigner = FacilitatorSvmSigner & {
  getAccountInfo: NonNullable<FacilitatorSvmSigner["getAccountInfo"]>;
  getLatestBlockhash: NonNullable<FacilitatorSvmSigner["getLatestBlockhash"]>;
  getSlot: NonNullable<FacilitatorSvmSigner["getSlot"]>;
  getSigner(feePayer: Address): FacilitatorSigningCapabilities;
};

const PAYMENT_CHANNEL_FACILITATOR_METHODS = [
  "getSigner",
  "getAccountInfo",
  "getLatestBlockhash",
  "getSlot",
] as const satisfies readonly (keyof PaymentChannelFacilitatorSigner)[];

/**
 * Assert a facilitator signer exposes every optional cap payment-channel
 * facilitator work needs.
 *
 * @param signer - Facilitator signer to validate
 * @param label - Component name for error messages
 * @throws Error when a required capability is missing
 */
export function assertPaymentChannelFacilitatorSigner(
  signer: FacilitatorSvmSigner,
  label: string,
): asserts signer is PaymentChannelFacilitatorSigner {
  for (const method of PAYMENT_CHANNEL_FACILITATOR_METHODS) {
    if (typeof signer[method] !== "function") {
      throw new Error(`${label} requires ${method} on the signer.`);
    }
  }
}

/**
 * Kit-compatible RPC adapter so generated account fetch helpers read through
 * the facilitator signer.
 *
 * @param signer - Payment-channel facilitator signer
 * @param network - CAIP-2 network identifier
 * @returns Minimal RPC surface for {@link fetchMaybeChannel}
 */
export function accountFetchRpc(
  signer: PaymentChannelFacilitatorSigner,
  network: string,
): Parameters<typeof fetchMaybeChannel>[0] {
  return {
    getAccountInfo: (
      accountAddress: Address,
      config?: { commitment?: string; encoding?: string },
    ) => ({
      send: async () => ({
        context: { slot: 0n },
        value: await signer.getAccountInfo(accountAddress.toString(), network, {
          commitment: config?.commitment,
          encoding: config?.encoding,
        }),
      }),
    }),
  } as Parameters<typeof fetchMaybeChannel>[0];
}
